lean/main/dsls.lean

```diff
@@ -0,0 +1,514 @@
+quote the string back into `Syntax`, and then we finally build
+the `AExp.Avar`.
+-/
+macro_rules
+| `([imp_aexp| $name:ident ]) => do
+   let nameStr : String := name.getId.toString
+   let nameStx : Syntax := Lean.quote nameStr
+   `(AExp.AVar $(nameStx))
+
+def eg_AExp_ident_macro: AExp := [imp_aexp|  foo]
+#reduce eg_AExp_ident_macro
+-- AExp.AVar "foo"
+
+/-
+In contrast to the `macro_rules` based solution, see that
+we build the expression node by hand, using lower level
+functions such as `mkAppM` to build a function application,
+and `mkStrLit: String → Expr` to convert a string into an `Expr`.
+-/
+def elab_aexp_ident (s: Syntax): TermElabM Expr :=
+match s with 
+| `(imp_aexp| $n:ident) =>
+      mkAppM ``AExp.AVar #[mkStrLit n.getId.toString]
+| _ => do 
+  dbg_trace "elab_aexp_ident failed."
+  throwUnsupportedSyntax
+
+/-
+We add our `elab` rule, which says that we can try to elaborate
+a `s:imp_aexp` with `elab_aexp_ident`. 
+-/
+elab "[imp_aexp'|" s:imp_aexp "]" : term => elab_aexp_ident s
+
+def eg_AExp_ident_elab: AExp := [imp_aexp'|  foo]
+#reduce eg_AExp_ident_elab
+-- AExp.AVar "foo"
+
+/-
+We test that our new elaboration rule did not interfere with 
+our previous rule to parse numbers:
+-/
+def eg_AExp_num2_elab: AExp := [imp_aexp'|  43]
+#reduce eg_AExp_num2_elab
+-- elab_aexp_ident failed.
+-- AExp.AVar "43"
+
+/-
+See that we have *not* lost the ability to parse numbers when we introduced
+our new `elab_aexp_ident` rule, even though the two declarations look identical
+on the left-hand-side of the rule:
+
+```
+elab "[imp_aexp'|" s:imp_aexp "]" : term => elab_aexp_num name
+elab "[imp_aexp'|" s:imp_aexp "]" : term => elab_aexp_ident name
+```
+
+This is because, as we discussed above, introducing new `elab` rules ensures
+that these rules are run in sequence, and this allows for the Lean syntax
+to be extended gracefully in an open-ended fashion.
+
+We can see from the output that Lean did try to runthe rule `elab_aexp_ident` which failed.
+It then fell back to running `elab_aexp_num`, which succeeded.
+-/
+
+/-
+We can try a piece of grammar that has not been handled yet, and see how both elaborators
+will be invoked in succession:
+-/
+
+def eg_AExp_fail: AExp := [imp_aexp'| 42 + 43]
+-- elab_aexp_ident failed.
+-- elab_aexp_num failed
+
+/-
+Clearly, both parses `elab_aexp_ident` and `elab_aexp_num`
+are tried in succession, and both fail, leading to an error.
+
+We shall fix this omission, and finally parse the addition node. 
+We go directly to parsing with `elab`, since the `macro_rules`
+approach does not have anything more interesting to say.
+-/
+
+def elab_aexp_plus (s: Syntax): TermElabM Expr := 
+match s with 
+| `(imp_aexp| $x:imp_aexp + $y:imp_aexp) => do 
+     -- recrsively expand xExpr, yExpr via `Term.elabTerm`
+     let xExpr <- Term.elabTerm (<- `([imp_aexp'| $x])) (expectedType? := none)
+     let yExpr <- Term.elabTerm (<- `([imp_aexp'| $y])) (expectedType? := none)
```

Comment:
Can we use `elab_AExp_num` here? I don't want to be reliant on `Term.elabTerm`

Notice that I didn't use `Term.elabTerm` to build the DSL [here](https://github.com/arthurpaulino/FxyLang/blob/master/MetaDebug.lean)

Reply:
I don't believe it's possible. calling `elab_AExp_num` will mean that when elaborating something like `1 + (2 + 3)`, we try to elaborate `yExpr = (2 + 3)` using `elab_AExp_num` which will not work.

Essentially, we use `Term.elabTerm` to give us the ability to recursively expand the _syntax category_ `imp_aexp`, whose elaboration is being defined piece-by-piece.